### PR TITLE
Config schema validation

### PR DESCRIPTION
**Pull request by live1206**
- Add schema validation for Config
- Log Config errors
- Fix some Config schema errors

### CODE REVIEW

File: docs/specs/config.yml [hunk 1]
```diff
@@ -325,10 +325,26 @@ inputs:
   docfx.yml: |
     redirections:
       docs/a.md: /absolute/path1
-    redirectionsWithoutId:
+    redirectionsWithoutDocumentId:
```
Comment (yufeih) [inline docs/specs/config.yml]: let's stick with `redirectionsWithoutId`, this feels like a breaking change.

File: src/docfx/config/Config.cs [hunk 2]
```diff
@@ -146,97 +146,111 @@ public static Config Load(string docsetPath, CommandLineOptions options, bool ex
         /// Load the config if it exists under <paramref name="docsetPath"/>
         /// </summary>
         /// <returns>Whether config exists under <paramref name="docsetPath"/></returns>
-        public static bool LoadIfExists(string docsetPath, CommandLineOptions options, out Config config, bool extend = true, RestoreMap restoreMap = null)
+        public static bool LoadIfExists(List<Error> errors, string docsetPath, CommandLineOptions options, out Config config, bool extend = true, RestoreMap restoreMap = null)
```
Comment (yufeih) [inline src/docfx/config/Config.cs]: How about `(errors, config) LoadIfExists(...)`? when it does not exist, return `(errors, null)`

File: src/docfx/config/Config.cs [hunk 3]
```diff
@@ -146,97 +146,111 @@ public static Config Load(string docsetPath, CommandLineOptions options, bool ex
...
-            ValidateLocale(config);
-            ValidateContributorConfig(config.Contribution, docsetPath);
+            var errors = new List<Error>();
+            errors.AddRange(ValidateLocale(config));
+            errors.AddRange(ValidateContributorConfig(config.Contribution, docsetPath));
+            return errors;
         }
 
-        private static void ValidateLocale(Config config)
+        private static List<Error> ValidateLocale(Config config)
```
Comment (yufeih) [inline src/docfx/config/Config.cs]: can this be implemented as an DataAnnotation extension, like `LocaleAttribute` ?

File: src/docfx/config/Config.cs [hunk 4]
```diff
@@ -146,97 +146,111 @@ public static Config Load(string docsetPath, CommandLineOptions options, bool ex
...
             catch (CultureNotFoundException e)
             {
-                throw Errors.InvalidLocale(config.Locale).ToException(e);
+                errors.Add(Errors.InvalidLocale(config.Locale));
             }
+            return errors;
         }
 
-        private static void ValidateContributorConfig(ContributionConfig config, string docsetPath)
+        private static List<Error> ValidateContributorConfig(ContributionConfig config, string docsetPath)
```
Comment (yufeih) [inline src/docfx/config/Config.cs]: can this be implemented as `ExternalFileAttribute` ?

File: src/docfx/restore/Restore.cs [hunk 5]
```diff
@@ -17,22 +17,24 @@ public static async Task Run(string docsetPath, CommandLineOptions options, Repo
         {
             // Restore has to use Config directly, it cannot depend on Docset,
             // because Docset assumes the repo to physically exist on disk.
+            var errors = new List<Error>();
```
Comment (yufeih) [inline src/docfx/restore/Restore.cs]: this is not thread safe if `RestoreDocset` runs in parallel.

File: src/docfx/config/Config.cs (hunk 2 above)
Comment (live1206) [reply src/docfx/config/Config.cs]: Thought about that as well, but we do use the bool return value to check if config file exists to control diff flows. So seems better to keep the signature and pass in errors list

---
In reply to: [211462329](https://github.com/dotnet/docfx/pull/3243#discussion_r211462329) [](ancestors = 211462329)

File: src/docfx/restore/Restore.cs (hunk 5 above)
Comment (live1206) [reply src/docfx/restore/Restore.cs]: Right, will update it to ConcurrentBag

---
In reply to: [211463029](https://github.com/dotnet/docfx/pull/3243#discussion_r211463029) [](ancestors = 211463029)

File: src/docfx/config/Config.cs (hunk 2 above)
Comment (yufeih) [reply src/docfx/config/Config.cs]: or if can be `bool TryLoad(out var (errors, config))`, just don't let it populate an errors list in an API boundary like here, it is easy to introduce mutli threading problems like below.

---
In reply to: [211463455](https://github.com/dotnet/docfx/pull/3243#discussion_r211463455) [](ancestors = 211463455,211462329)

Comment (live1206) [reply src/docfx/config/Config.cs]: ok, sounds better

---
In reply to: [211463951](https://github.com/dotnet/docfx/pull/3243#discussion_r211463951) [](ancestors = 211463951,211463455,211462329)

File: src/docfx/config/Config.cs (hunk 4 above)
Comment (live1206) [reply src/docfx/config/Config.cs]: Will create a new PR for this

---
In reply to: [211462828](https://github.com/dotnet/docfx/pull/3243#discussion_r211462828) [](ancestors = 211462828)

File: src/docfx/config/Config.cs (hunk 3 above)
Comment (live1206) [reply src/docfx/config/Config.cs]: Will create a separate PR for this

---
In reply to: [211462516](https://github.com/dotnet/docfx/pull/3243#discussion_r211462516) [](ancestors = 211462516)

File: src/docfx/restore/Restore.cs (hunk 5 above)
Comment (live1206) [reply src/docfx/restore/Restore.cs]: Pass Report in RestoreDocset instead

---
In reply to: [211463509](https://github.com/dotnet/docfx/pull/3243#discussion_r211463509) [](ancestors = 211463509,211463029)

**Pull request merged**